torch/csrc/jit/codegen/cuda/expr_simplifier.cpp

```diff
@@ -0,0 +1,430 @@
+  for (auto v : variables) {
+    if (v->isOneOf<TensorView, kir::TensorIndex>()) {
+      return true;
+    }
+  }
+  return false;
+}
+
+} // namespace
+
+namespace assoc_comm_reordering {
+
+// Note: [Reordering associative and commutative operators]
+//
+// For binary operators that is both associative and commutative, we can freely
+// change the order of operands and add/remove parenthesis without changing the
+// result. For example, + is both associative and commutative, so we have:
+// a + b + c := (a + b) + c = a + (b + c) = (b + a) + c = b + (a + c) = ...
+// For these operators, the most convenient way for handling them is to flatten
+// them. For example, for the above a + b + c, all we need to know is we are
+// adding these three variables together. We don't really care whether we are
+// adding a and b first, or adding a and c first, or whether we are adding a to
+// c or adding c to a. `FlattenedAssocCommOp` is the class that represents this
+// flattened perspective.
+//
+// The reordering of associative and commutative operators is mostly useful for
+// index hoisting. For example, if I have a loop structure and index:
+//   FOR i1
+//     FOR i2
+//       FOR i3
+//         index = ((i3 + i2) + i1) + 256
+// There is no hoisting opportunity for this index in this loop structure.
+// However, if I transform the index into index = ((256 + i1) + i2) + i3,
+// then I can hoist the index as
+//   FOR i1
+//     i4 = (256 + i1)
+//     FOR i2
+//       i5 = i4 + i2
+//       FOR i3
+//         index = i5 + i3
+// This minimizes the total number of computations.
+
+bool isAssociativeAndCommutative(BinaryOpType type) {
+  return type == BinaryOpType::Add || type == BinaryOpType::Mul ||
+      type == BinaryOpType::And || type == BinaryOpType::Or ||
+      type == BinaryOpType::Xor;
+}
+
+// The expression type that represents the flattened ops. For example, if I have
+// out = a + b + 3 + c + 5, then I will have:
+//   FlattenedAssocCommOp {
+//     inputs: [a, b, c]
+//     putputs: [out]
+//     constant term: 8
+//   }
+//
+// TODO: refactor IR printing so that private exprs like this can be printed.
+// This will be very helpful for debugging.
+class FlattenedAssocCommOp : public Expr {
+ public:
+  using Expr::Expr;
+
+  FlattenedAssocCommOp(
+      IrBuilderPasskey passkey,
+      BinaryOpType op,
+      Val* out,
+      std::vector<Val*> terms)
+      : Expr(passkey) {
+    TORCH_CHECK(
+        isAssociativeAndCommutative(op),
+        "Can only flatten associative and commutative ops");
+    addAttribute(
+        IrBuilder::create<Attribute<BinaryOpType>>(passkey.ir_container_, op));
+    addOutput(out);
+    for (auto v : terms) {
+      // Note that `addInput` is overriden in this class.
+      addInput(v);
+    }
+  }
+
+  NVFUSER_DECLARE_CLONE_AND_CREATE
+
+  virtual const char* getOpString() const override {
+    return "FlattenedAssocCommOp";
+  }
+
+  DataType dtype() const {
+    return *output(0)->getDataType();
```

Comment:
Is there an assertion on `hasSimilarType` for all the inputs?

Reply:
Yes, it is inside the overriden `addInput`